File: app/src/main/java/acr/browser/lightning/database/history/HistoryDatabase.java
/*
 * Copyright 2014 A.C.R. Development
 */
package acr.browser.lightning.database.history;

import android.app.Application;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.DatabaseUtils;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import com.anthonycr.bonsai.Completable;
import com.anthonycr.bonsai.CompletableAction;
import com.anthonycr.bonsai.CompletableSubscriber;
import com.anthonycr.bonsai.Single;
import com.anthonycr.bonsai.SingleAction;
import com.anthonycr.bonsai.SingleSubscriber;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import acr.browser.lightning.R;
import acr.browser.lightning.database.HistoryItem;


/**
 * The disk backed download database.
 * See {@link HistoryModel} for method
 * documentation.
 */
@Singleton
@WorkerThread
public class HistoryDatabase extends SQLiteOpenHelper implements HistoryModel {

    // All Static variables
    // Database Version
    private static final int DATABASE_VERSION = 2;

    // Database Name
    private static final String DATABASE_NAME = "historyManager";

    // HistoryItems table name
    private static final String TABLE_HISTORY = "history";

    // HistoryItems Table Columns names
    private static final String KEY_ID = "id";
    private static final String KEY_URL = "url";
    private static final String KEY_TITLE = "title";
    private static final String KEY_TIME_VISITED = "time";

    @Nullable private SQLiteDatabase mDatabase;

    @Inject
    public HistoryDatabase(@NonNull Application application) {
        super(application, DATABASE_NAME, null, DATABASE_VERSION);
    }

    // Creating Tables
    @Override
    public void onCreate(@NonNull SQLiteDatabase db) {
        String CREATE_HISTORY_TABLE = "CREATE TABLE " + TABLE_HISTORY + '(' + KEY_ID
            + " INTEGER PRIMARY KEY," + KEY_URL + " TEXT," + KEY_TITLE + " TEXT,"
            + KEY_TIME_VISITED + " INTEGER" + ')';
        db.execSQL(CREATE_HISTORY_TABLE);
    }

    // Upgrading database
    @Override
    public void onUpgrade(@NonNull SQLiteDatabase db, int oldVersion, int newVersion) {
        // Drop older table if it exists
        db.execSQL("DROP TABLE IF EXISTS " + TABLE_HISTORY);
        // Create tables again
        onCreate(db);
    }

    @NonNull
    private static HistoryItem fromCursor(@NonNull Cursor cursor) {
        HistoryItem historyItem = new HistoryItem();
        historyItem.setUrl(cursor.getString(1));
        historyItem.setTitle(cursor.getString(2));
        historyItem.setImageId(R.drawable.ic_history);

        return historyItem;
    }

    @WorkerThread
    @NonNull
    private synchronized SQLiteDatabase lazyDatabase() {
        if (mDatabase == null || !mDatabase.isOpen()) {
            mDatabase = this.getWritableDatabase();
        }
        return mDatabase;
    }

    @NonNull
    @Override
    public Completable deleteHistory() {
        return Completable.create(new CompletableAction() {
            @Override
            public void onSubscribe(@NonNull CompletableSubscriber subscriber) {
                lazyDatabase().delete(TABLE_HISTORY, null, null);
                lazyDatabase().close();

                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Completable deleteHistoryItem(@NonNull final String url) {
        return Completable.create(new CompletableAction() {
            @Override
            public void onSubscribe(@NonNull CompletableSubscriber subscriber) {
                lazyDatabase().delete(TABLE_HISTORY, KEY_URL + " = ?", new String[]{url});

                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Completable visitHistoryItem(@NonNull final String url, @Nullable final String title) {
        return Completable.create(new CompletableAction() {
            @Override
            public void onSubscribe(@NonNull CompletableSubscriber subscriber) {
                ContentValues values = new ContentValues();
                values.put(KEY_TITLE, title == null ? "" : title);
                values.put(KEY_TIME_VISITED, System.currentTimeMillis());

                Cursor cursor = lazyDatabase().query(false, TABLE_HISTORY, new String[]{KEY_URL},
                    KEY_URL + " = ?", new String[]{url}, null, null, null, "1");

                if (cursor.getCount() > 0) {
                    lazyDatabase().update(TABLE_HISTORY, values, KEY_URL + " = ?", new String[]{url});
                } else {
                    addHistoryItem(new HistoryItem(url, title == null ? "" : title));
                }

                cursor.close();
            }
        });
    }

    @NonNull
    @Override
    public Single<List<HistoryItem>> findHistoryItemsContaining(@NonNull final String query) {
        return Single.create(new SingleAction<List<HistoryItem>>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<List<HistoryItem>> subscriber) {
                List<HistoryItem> itemList = new ArrayList<>(5);

                String search = '%' + query + '%';

                Cursor cursor = lazyDatabase().query(TABLE_HISTORY, null, KEY_TITLE + " LIKE ? OR " + KEY_URL + " LIKE ?",
                    new String[]{search, search}, null, null, KEY_TIME_VISITED + " DESC", "5");

                while (cursor.moveToNext()) {
                    itemList.add(fromCursor(cursor));
                }

                cursor.close();

                subscriber.onItem(itemList);
                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Single<List<HistoryItem>> lastHundredVisitedHistoryItems() {
        return Single.create(new SingleAction<List<HistoryItem>>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<List<HistoryItem>> subscriber) {
                List<HistoryItem> itemList = new ArrayList<>(100);
                Cursor cursor = lazyDatabase().query(TABLE_HISTORY, null, null, null, null, null, KEY_TIME_VISITED + " DESC", "100");

                while (cursor.moveToNext()) {
                    itemList.add(fromCursor(cursor));
                }

                cursor.close();

                subscriber.onItem(itemList);
                subscriber.onComplete();
            }
        });
    }

    @WorkerThread
    private synchronized void addHistoryItem(@NonNull HistoryItem item) {
        ContentValues values = new ContentValues();
        values.put(KEY_URL, item.getUrl());
        values.put(KEY_TITLE, item.getTitle());
        values.put(KEY_TIME_VISITED, System.currentTimeMillis());
        lazyDatabase().insert(TABLE_HISTORY, null, values);
    }

    @WorkerThread
    @Nullable
    synchronized String getHistoryItem(@NonNull String url) {
        Cursor cursor = lazyDatabase().query(TABLE_HISTORY, new String[]{KEY_ID, KEY_URL, KEY_TITLE},
            KEY_URL + " = ?", new String[]{url}, null, null, null, "1");
        String m = null;
        if (cursor != null) {
            cursor.moveToFirst();
            m = cursor.getString(0);

            cursor.close();
        }
        return m;
    }

    @WorkerThread
    @NonNull
    synchronized List<HistoryItem> getAllHistoryItems() {
        List<HistoryItem> itemList = new ArrayList<>();

        Cursor cursor = lazyDatabase().query(TABLE_HISTORY, null, null, null, null, null, KEY_TIME_VISITED + " DESC");

        while (cursor.moveToNext()) {
            itemList.add(fromCursor(cursor));
        }

        cursor.close();

        return itemList;
    }

    @WorkerThread
    synchronized long getHistoryItemsCount() {
        return DatabaseUtils.queryNumEntries(mDatabase, TABLE_HISTORY);
    }
}

File: app/src/main/java/acr/browser/lightning/database/history/HistoryDatabase.kt
/*
 * Copyright 2014 A.C.R. Development
 */
package acr.browser.lightning.database.history

import acr.browser.lightning.database.HistoryEntry
import acr.browser.lightning.database.databaseDelegate
import acr.browser.lightning.extensions.firstOrNullMap
import acr.browser.lightning.extensions.useMap
import android.app.Application
import android.content.ContentValues
import android.database.Cursor
import android.database.DatabaseUtils
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import androidx.annotation.WorkerThread
import io.reactivex.Completable
import io.reactivex.Single
import javax.inject.Inject
import javax.inject.Singleton


/**
 * The disk backed download database. See [HistoryRepository] for function documentation.
 */
@Singleton
@WorkerThread
class HistoryDatabase @Inject constructor(
    application: Application
) : SQLiteOpenHelper(application, DATABASE_NAME, null, DATABASE_VERSION), HistoryRepository {

    private val database: SQLiteDatabase by databaseDelegate()

    // Creating Tables
    override fun onCreate(db: SQLiteDatabase) {
        val createHistoryTable = "CREATE TABLE $TABLE_HISTORY(" +
            " $KEY_ID INTEGER PRIMARY KEY," +
            " $KEY_URL TEXT," +
            " $KEY_TITLE TEXT," +
            " $KEY_TIME_VISITED INTEGER" +
            ")"
        db.execSQL(createHistoryTable)
    }

    // Upgrading database
    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Drop older table if it exists
        db.execSQL("DROP TABLE IF EXISTS $TABLE_HISTORY")
        // Create tables again
        onCreate(db)
    }

    override fun deleteHistory(): Completable = Completable.fromAction {
        database.run {
            delete(TABLE_HISTORY, null, null)
            close()
        }
    }

    override fun deleteHistoryEntry(url: String): Completable = Completable.fromAction {
        database.delete(TABLE_HISTORY, "$KEY_URL = ?", arrayOf(url))
    }

    override fun visitHistoryEntry(url: String, title: String?): Completable =
        Completable.fromAction {
            val values = ContentValues().apply {
                put(KEY_TITLE, title ?: "")
                put(KEY_TIME_VISITED, System.currentTimeMillis())
            }

            database.query(
                false,
                TABLE_HISTORY,
                arrayOf(KEY_ID),
                "$KEY_URL = ?",
                arrayOf(url),
                null,
                null,
                null,
                "1"
            ).use {
                if (it.count > 0) {
                    database.update(TABLE_HISTORY, values, "$KEY_URL = ?", arrayOf(url))
                } else {
                    addHistoryEntry(HistoryEntry(url, title ?: ""))
                }
            }
        }

    override fun findHistoryEntriesContaining(query: String): Single<List<HistoryEntry>> =
        Single.fromCallable {
            val search = "%$query%"

            return@fromCallable database.query(
                TABLE_HISTORY,
                null,
                "$KEY_TITLE LIKE ? OR $KEY_URL LIKE ?",
                arrayOf(search, search),
                null,
                null,
                "$KEY_TIME_VISITED DESC",
                "5"
            ).useMap { it.bindToHistoryEntry() }
        }

    override fun lastHundredVisitedHistoryEntries(): Single<List<HistoryEntry>> =
        Single.fromCallable {
            database.query(
                TABLE_HISTORY,
                null,
                null,
                null,
                null,
                null,
                "$KEY_TIME_VISITED DESC",
                "100"
            ).useMap { it.bindToHistoryEntry() }
        }

    @WorkerThread
    private fun addHistoryEntry(item: HistoryEntry) {
        database.insert(TABLE_HISTORY, null, item.toContentValues())
    }

    @WorkerThread
    fun getHistoryEntry(url: String): String? =
        database.query(
            TABLE_HISTORY,
            arrayOf(KEY_ID, KEY_URL, KEY_TITLE),
            "$KEY_URL = ?",
            arrayOf(url),
            null,
            null,
            null,
            "1"
        ).firstOrNullMap { it.getString(0) }


    fun getAllHistoryEntries(): List<HistoryEntry> {
        return database.query(
            TABLE_HISTORY,
            null,
            null,
            null,
            null,
            null,
            "$KEY_TIME_VISITED DESC"
        ).useMap { it.bindToHistoryEntry() }
    }

    fun getHistoryEntriesCount(): Long = DatabaseUtils.queryNumEntries(database, TABLE_HISTORY)

    private fun HistoryEntry.toContentValues() = ContentValues().apply {
        put(KEY_URL, url)
        put(KEY_TITLE, title)
        put(KEY_TIME_VISITED, lastTimeVisited)
    }

    private fun Cursor.bindToHistoryEntry() = HistoryEntry(
        url = getString(1),
        title = getString(2),
        lastTimeVisited = getLong(3)
    )

    companion object {

        // Database version
        private const val DATABASE_VERSION = 2

        // Database name
        private const val DATABASE_NAME = "historyManager"

        // HistoryEntry table name
        private const val TABLE_HISTORY = "history"

        // HistoryEntry table columns names
        private const val KEY_ID = "id"
        private const val KEY_URL = "url"
        private const val KEY_TITLE = "title"
        private const val KEY_TIME_VISITED = "time"

    }
}
